### DISCUSSION TITLE
Treat Queue and current Playlist/Radio separately like Spotify

### DISCUSSION THREAD

**Discussion by Auram** [music-assistant/support#2895] [Feature requests and ideas]
So in Spotify when you select a playlist or radio and set it to shuffle or play sequentially you can still add things to the top of the queue in the order they were added or, in the case of an album, in their sequential order. This is helpful if you're just listening to something on shuffle, a song from an album you like comes on, and you want to hear that whole album in order but then automatically go back to the previous playlist.

Currently in MA if you try this with Play Next or Add to Queue with shuffle enabled it will just add the album's tracks and then shuffle them, which is not desired in most cases I'd say. At least that's how I feel as someone who has used the Spotify app for ages.

![Screenshot_20240910_094505_Spotify](https://github.com/user-attachments/assets/811d0178-eede-4264-b44b-b228c0cafa64)
In this screenshot here you can see how I've queued up an album which is playing non-shuffled and then once it's done it will just revert back to the playlist I was shuffling beforehand.

**Comment by OzGav** [comment]
Turn shuffle off before you add the new items

**Reply by Auram** [reply]
That is not a solution as I'd want it to go back to shuffling after the new added items play in order.

**Reply by OzGav** [reply]
Then you could turn shuffle back on

**Reply by Auram** [reply]
Yes, but that's a lot of babysitting that this feature wouldn't need. MA already has way more clicks to do things then you need in Spotify natively, I don't wanna go adding more.

**Comment by dvd0bvb** [comment]
Consider this flow:
I have guests and I let them add songs to the queue. I don't want to have to keep manually adding songs when the queue empties, I want to hang out so I enable radio mode. Alice wants to pick a song, so she adds it to the queue. No one else has picked a song yet so her song should play next. Bob picks a song after Alice has added hers to the queue so Bob's song should play directly after Alice's. When Bob's song is finished the next radio track should play until someone else picks a song.

Currently Bob's options are to step on Alice's toes and hit "play next", manually move his track (lol), skip tracks (which are potentially other people's picks), or wait for however many tracks are left. Separating manually selected and radio tracks would solve this as both Alice and Bob could just hit "add to queue".

**Comment by Kreinoee** [comment]
I think this is somewhat related to this feature request: #5240 in where I posted some arguments relevant to this, so I will repeat them here:

First of, I fully agree that tying the shuffle feature to the action of adding stuff to the queue, instead of as a setting on the queue/player itself, is much more flexible when it comes to use cases. I written that spotify do it like that, but I can add that the same goes for Qobuz, and it's a much nicer user experience that what MA offer currently. 

But beside the increased flexibility, the current approach makes the UI around the add to queue operations unoptimal, as they behaves differently depending on the shuffle state—a state that you may or may not be aware of when adding items to the queue— hence it feels inconsistent. Some of the add features' behavior are, at least in my mind, also confusing. I'll try to walk through them here so I can better describe what I mean by the add items actions being inconsistent:

## Add to the queue
Shuffle off → Add item(s) to the end of the queue
Shuffle on → Add item(s) to random places in the queue

So this is a simple example of what I mean by inconsistent: it does two different things, and you need to be aware of a separate state to know what your action will do. But keep reading, because this is not the only problem.

## Play now (keep queue)
I would like to separate this into two different types of adds, as it behaves differently when adding one or multiple items, at least when shuffle is on.

### Add single item
Shuffle off → Add item after the currently playing item in the queue, and skip the queue one forward so that it immediately starts playing the first of the items just added
Shuffle on → Does the same as if shuffle is off.

So in this case there is no inconsistency.

### Add multiple items
Shuffle off → All items are added after the currently playing item in the queue, and the queue is skipped one item forward.
Shuffle on → This one behaves differently in stable and beta. In stable, it adds one of the items just after the currently playing one, but the rest of the items are added at random places in the queue. It then skips the queue one item forward, so that it immediately plays one of the added items. In the beta, it adds all of them randomly, then starts playing a completely different item from the queue. That looks like a bug.

This case is, to me, inconsistent in two ways:

1. It does different things depending on whether shuffle is on or off (just like "Add to queue").
2. It's inconsistent in its treatment of the files you add. You're adding all of them, telling MA to play them now, but it only plays one of them now, and puts the others in at random places. In my opinion, not what you would expect.

## Play next (keep queue)
Almost the same as "Play now (keep queue)", but it doesn't skip the queue one forward, so it keeps playing the current song. However, there is one difference: when adding multiple items, it seems in the beta like it always takes the first item added (not one at random), and adds it as the next in the queue. But just as with play now when adding multiple files, it does not exactly do what you would expect of it.

## The rest
There are more add options, but I think this should be enough to illustrate why I call them inconsistent.

My suggestion is that you drop the shuffle setting as a state in general, so that it's neither a setting on the player (as in current stable) nor the queue (as in beta). Instead, implement more "add to queue" options like:

* **Shuffle into queue** → Adds the item(s) at random places in the queue. Should be present for both single items and multi-item adds.
* **Add shuffled to end of queue** → Shuffles the added items and adds the result at the end of the queue. Should be present only for multi-item adds, as for a single item it would be the same as a plain "Add to queue".

If that is done, the existing add features would stop having different—and sometimes confusing—behavior depending on whether shuffle is on or off, and they would now always do exactly what their name says. And shuffle is still possible with the new add options, and would support all the use cases that this feature request is about. Both the queue and the add operations thereby becomes more "what you se if what you get" like.  It also solves the  "shuffle bleeds into later playback" problem stated by #5240. Finally without knowing the code base, I could even imaging that it would make it simpler, as there is one state less to care about, but that is very much an assumption :crossed_fingers:
